More cross refs and doctests for linalg

kshyatt opened the pull request:
And fixed some failing doctests

kshyatt committed:
Move eigen docs inline, add cross refs

kshyatt committed:
Add cross refs to sparse linalg docs

kshyatt committed:
More cross refs for cholesky docs

kshyatt committed:
Cross references for Schur docs

kshyatt committed:
Fix broken doctest for symmetric/Hermitian

kshyatt committed:
More references in arnoldi

kshyatt committed:
More doctest examples for BLAS, some cleanup

andreasnoack commented:
LGTM

Sacha0 commented on base/linalg/blas.jl:
```diff
@@ -799,7 +858,8 @@ end
 `tA` (transpose `A`) and `ul` (triangle of `A` used). `dA` indicates if `A` is
-unit-triangular (the diagonal is assumed to be all ones). Returns the updated `b`.
+unit-triangular (the diagonal is assumed to be all ones if `dA` is `U`, and assumed to have
+non-unit values if `dA` is `N`). Returns the updated `b`.
```
Should the ``` `U` ``` be ```  `'U'` ```? (Likewise ``` `N` ```.)

kshyatt replied:
Yeah! Thanks for catching this.

Sacha0 commented on base/linalg/blas.jl:
```diff
@@ -1116,16 +1178,16 @@ end
 Rank-k update of the symmetric matrix `C` as `alpha*A*A.' + beta*C` or `alpha*A.'*A +
-beta*C` according to whether `trans` is 'N' or 'T'. When `uplo` is 'U' the upper triangle of
-`C` is updated ('L' for lower triangle). Returns `C`.
+beta*C` according to whether `trans` is 'N' or 'T'. When `uplo` is `'U'` the upper triangle of
```
Likewise here, should `'N'` and `'T'` be ``` `'N'` ``` and ``` `'T'` ``` or as they are?

kshyatt replied:
I think we want both sets of ticks.

Sacha0 commented on base/linalg/schur.jl:
```diff
@@ -102,7 +102,7 @@ end
     ordschur!(F::Schur, select::Union{Vector{Bool},BitVector}) -> F::Schur
 
-Same as `ordschur` but overwrites the factorization `F`.
+    Same as [`ordschur`](@ref) but overwrites the factorization `F`.
```
Indentation intended?

Sacha0 commented on base/sparse/umfpack.jl:
```diff
@@ -131,7 +131,7 @@ The relation between `F` and `A` is
 - [`det`](@ref)
 
-** Implementation note **
+!!! note
```
Did you intend to nix the `Implementation`?

Sacha0 replied:
Either way on this end :).

andreasnoack replied:
I think `note` is fine.

MichaelHatherly commented on base/linalg/arnoldi.jl:
````diff
@@ -349,9 +349,9 @@ X = sprand(10, 5, 0.2)
 ```
 
-**Implementation note**
+!!! Implementation note
````
Should be `!!! note "Implementation"` if you'd like a `note` admonition with the title "Implementation".

MichaelHatherly commented on base/linalg/blas.jl:
```diff
@@ -217,20 +217,41 @@ scal(n, DA, DX, incx) = scal!(n, DA, copy(DX), incx)
 Dot product of two vectors consisting of `n` elements of array `X` with stride `incx` and
 `n` elements of array `Y` with stride `incy`.
+
+**Example**:
```
The doc style guide suggests using `# Examples`, though I don't think it's being followed that widely in base yet; perhaps due to how headers are currently rendered in the REPL (they take up a bit to much vertical space I think).

MichaelHatherly commented on base/linalg/eigen.jl:
```diff
@@ -170,6 +170,22 @@ function eigvals!{T<:BlasComplex}(A::StridedMatrix{T}; permute::Bool=true, scale
 end
+
+"""
+
```
Extra newline here.

kshyatt replied:
Thanks for the catch!

tkelman commented on base/linalg/blas.jl:
````diff
@@ -339,6 +369,15 @@ nrm2(x::Array) = nrm2(length(x), pointer(x), 1)
+# Example:
+```jldoctest
+julia> Base.BLAS.asum(10, im*ones(10), 2)
+5.0
````
is this reading past the end of the array or what?

kshyatt replied:
Oh shoot, it might be. I'll fix it.

tkelman replied:
is the stopping criteria set by blas or the julia wrapper?

kshyatt replied:
We just pass `n` in to BLAS, so I'm guessing the former.

tkelman replied:
but we don't pass the length of the array, so how could it know to stop after 5 elements? http://www.netlib.org/lapack/explore-html/de/d05/dasum_8f_source.html

kshyatt replied:
I think what it's doing is saying "the total length of the array is 10, within those ten, take steps of width 2". For example:

```julia
julia> a = [1; 2; 4; 6; 8; 10]
6-element Array{Int64,1}:
  1
  2
  4
  6
  8
 10

julia> Base.BLAS.asum(6, float(a), 1)
31.0

julia> Base.BLAS.asum(6, float(a), 2)
13.0

julia> Base.BLAS.asum(6, float(a), 4)
9.0
```
It's taking elements up to the max length of `6`.

tkelman replied:
If you make `a` longer though, you'll get results that read from past the 6th element

kshyatt replied:
Are you sure?
```julia
julia> a = [1; 2; 4; 6; 8; 10];

julia> Base.BLAS.asum(4, float(a), 2)
13.0

julia> Base.BLAS.asum(4, float(a), 6)
1.0
```
Case 2 here suggests that's not what happens.

tkelman replied:
if the stride is 6, try with a 7-element vector
```
julia> a = [1; 2; 4; 8; 16; 32; 64];

julia> Base.BLAS.asum(4, float(a), 6)
65.0
```

kshyatt replied:
Sorry, I'm really dumb - is this a bug on our end? In BLAS? In my docs? All three?

tkelman replied:
It's hard to notice since I think we frequently get lucky that the space past the end of the array is happening to be set to zero. In C you might see nondeterministic results more often. Or you could be overly clever and take advantage of this intentionally with some subarray schemes.

I think the conclusion here is that BLAS functions don't do a whole lot of bounds checking if you haven't passed the array length. This potential issue may or may not be worth mentioning in the Julia docs, dunno. We should change any examples we give so that we're only reading from within the bounds of the arrays we create.

tkelman commented on base/linalg/blas.jl:
```diff
@@ -808,7 +869,8 @@ function trsv! end
 Returns the solution to `A*x = b` or one of the other two variants determined by `tA`
 (transpose `A`) and `ul` (triangle of `A` is used.) `dA` indicates if `A` is unit-triangular
-(the diagonal is assumed to be all ones).
+(the diagonal is assumed to be all ones if `dA` is `U`, and assumed to have non-unit values if
```
not really, it just reads the actual values from `A` in the non-unit case, right? they could still happen to equal 1 sometimes

tkelman commented on base/linalg/eigen.jl:
```diff
@@ -170,6 +170,21 @@ function eigvals!{T<:BlasComplex}(A::StridedMatrix{T}; permute::Bool=true, scale
+For general non-symmetric matrices it is possible to specify how the matrix is balanced
+before the eigenvector calculation. The option `permute=true` permutes the matrix to
+become closer to upper triangular, and `scale=true` scales the matrix by its diagonal
+elements to make rows and columns moreequal in norm. The default is `true` for both
```
more equal is 2 words

the signature above doesn't list the keywords

kshyatt committed:
Split up eigenfactorization docs

Make the docstrings accurately reflect the function signatures.
Add doctests for the symmetric cases.

kshyatt commented:
I got so mad at the documentation for eigenfunctions that I split it up into many pieces. @andreasnoack should probably review this again.

tkelman commented on base/linalg/blas.jl:
```diff
@@ -798,8 +857,10 @@ end
+`tA` (transpose `A` if `'T'`, `A` if `'N'`, conjugate-transpose if `'C'`)
+and `ul` (triangle of `A` used, `'U'` for upper, `'L`' for lower).
+`dA` indicates if `A` is unit-triangular (the diagonal is assumed to be all ones if `dA` is
+`'U'`, and assumed to have non-unit values if `dA` is `'N'`). Returns the updated `b`.
```
no, there's no assumption about the diagonal values at all in this case - they get read from

tkelman commented on base/linalg/blas.jl:
```diff
@@ -808,7 +869,8 @@ function trsv! end
 (transpose `A`) and `ul` (triangle of `A` is used.) `dA` indicates if `A` is unit-triangular
-(the diagonal is assumed to be all ones).
+(the diagonal is assumed to be all ones if `dA` is `U`, and this is not assumed if
+`dA` is `N` (so that the diagonal values are read in)).
```
instead of nested parens, how about "or the diagonal values are read from `A` if `dA` is `N`" ?

kshyatt replied:
Yeah, that's better

tkelman replied:
and should we make it clear that these are chars? `'U'` etc ?

kshyatt replied:
On that as well :)

tkelman commented on base/linalg/eigen.jl:
```diff
@@ -106,11 +100,15 @@ function eig(A::Union{Number, StridedMatrix}; permute::Bool=true, scale::Bool=tr
 Computes eigenvalues (`D`) and eigenvectors (`V`) of `A`.
 See [`eigfact`](@ref) for details on the
 `irange`, `vl`, and `vu` arguments
+(for [`SymTridigonal`](@ref), `Hermitian`, and
```
SymTridiagonal

kshyatt replied:
Typing is really hard

kshyatt commented:
I'm confused about why this is failing CI, @MichaelHatherly. I can run `make docs` and the doctests pass locally for me (Mac).

MichaelHatherly commented:
> I'm confused about why this is failing CI

Just download failure caused by https://github.com/JuliaLang/julia/pull/18588#discussion_r93816369, hopefully fixed by https://github.com/JuliaLang/julia/pull/19725.

tkelman reviewed (dismissed):
example for asum should not read past end of array

tkelman commented on base/linalg/blas.jl:
```diff
@@ -496,18 +549,18 @@ gemv(tA, A, x)
-Update vector `y` as `alpha*A*x + beta*y` or `alpha*A'*x + beta*y` according to `trans` ('N'
-or 'T'). The matrix `A` is a general band matrix of dimension `m` by `size(A,2)` with `kl`
-sub-diagonals and `ku` super-diagonals. Returns the updated `y`.
+Update vector `y` as `alpha*A*x + beta*y` or `alpha*A'*x + beta*y` according to `trans` (`'N'` (meaning not transposed)
```
I'd avoid the nested parentheses here

kshyatt replied:
Really what I think we should do is just put a big damn table of what all these chars mean at the top of the `BLAS` section

tkelman replied:
we should probably cross-reference that table in the relevant docstrings

kshyatt replied:
You mean for REPL users? Can do.

tkelman replied:
or anyone reading the stdlib side of the docs

tkelman commented on base/linalg/eigen.jl:
```diff
@@ -170,6 +168,18 @@ function eigvals!{T<:BlasComplex}(A::StridedMatrix{T}; permute::Bool=true, scale
+Returns the eigenvalues of `A`.
+
+For general non-symmetric matrices it is possible to specify how the matrix is balanced
+before the eigenvector calculation. The option `permute=true` permutes the matrix to
```
would it be more specific to say "eigenvalue calculation" here?

tkelman commented on base/linalg/symmetric.jl:
```diff
@@ -278,22 +278,106 @@ eigfact{T1<:Real,T2}(A::RealHermSymComplexHerm{T1,T2}) = (T = eltype(A); S = pro
+
+The following functions are available for `Eigen` objects: [`inv`](@ref), [`det`](@ref), and [`isposdef`](@ref).
+
+`vl` is the lower bound of the window of eigenvalues to search for, and `vu` is the upper bound.
```
so does this return a partial decomposition or ?

kshyatt replied:
@andreasnoack ?

andreasnoack replied:
Yes. The factorization is partial when either `vl/vu` or `irange` are used. I guess it won't really surprise anybody who uses these options but it might still be useful to state explicitly in the documentation.

andreasnoack replied:
Thinking a little longer, this is of course a problem for the methods the defined for `Eigen`. They'd have to check that the factorization is not truncated such that `det(F)==0` for a truncated factorization (i.e. when `size(F.vectors,2) < size(F.vectors,1)`) and `inv` should throw.

tkelman replied:
sounds like truncated factorizations should probably be a different type, and obtained via different method names? otherwise the result can't correctly be used in generic code

kshyatt replied:
Done!

tkelman commented on base/linalg/tridiag.jl:
```diff
@@ -202,6 +202,45 @@ eigmin(A::SymTridiagonal) = eigvals(A, 1:1)[1]
+    eigvecs(A::SymTridiagonal[, eigvals]) -> Matrix
+
+Returns a matrix `M` whose columns are the eigenvectors of `A`. (The `k`th eigenvector can
+be obtained from the slice `M[:, k]`.) The `permute` and `scale` keywords are the same as
```
doesn't looks like `permute` or `scale` work for `SymTridiagonal`

kshyatt committed:
Make a character table for BLAS, cleanup

tkelman commented on doc/src/stdlib/linalg.md:
```diff
@@ -157,6 +157,38 @@ linear algebra routines it is useful to call the BLAS functions directly.
+#### [Unit Diagonal](@id stdlib-blas-diag)
+| `diag`/`dX` | Meaning                                                   |
+|:------------|:----------------------------------------------------------|
+| `'N'`       | The diagonal values of the matrix `X` will be read.       | 
```
whitespace failure here

kshyatt replied:
D: but i ran `make docs`...

tkelman replied:
which I guess somehow doesn't depend on check-whitespace?

tkelman commented on base/linalg/blas.jl:
```diff
@@ -1044,31 +1105,27 @@ end
 symm(side, ul, A, B)
 
 """
-    symm(tA, tB, alpha, A, B)
```
where did this go?

kshyatt replied:
I killed it because it shouldn't exist, there is no such `symm` method.

tkelman replied:
hm. doc added in #3906, has it never been present and no one noticed until now?

kshyatt replied:
It doesn't really make sense for the _symmetric_ matrix multiply to take two transpose args. Also, it's not present in the Netlib docs.

tkelman commented on base/linalg/blas.jl:
```diff
@@ -1301,38 +1361,42 @@ end
+Update `B` as `alpha*A*B` or one of the other three variants determined by
+[`side`](@ref stdlib-blas-side) and [`tA`](@ref stdlib-blas-trans).
+Only the [`ul`](@ref stdlib-blas-uplo) triangle of `A` is used.
+The diagonal is assumed to be all ones (or not) according to [`dA`](@ref stdlib-blas-diag).
```
I like the wording you have in a few others of ```[`dA`](@ref stdlib-blas-diag) determines if the diagonal values are read or
are assumed to be all ones.```

kshyatt replied:
Got 'em :)

MichaelHatherly commented on base/linalg/blas.jl:
```diff
@@ -472,22 +524,24 @@ end
     gemv(tA, alpha, A, x)
 
-Returns `alpha*A*x` or `alpha*A'x` according to `tA` (transpose `A`).
+Returns `alpha*A*x` or `alpha*A'x` according to [`tA`](@ref stdlib-blas-trans).
```
`` [`tA`](@ref stdlib-blas-trans) `` is going to cross reference the `-` docstring rather than `@id stdlib-blas-trans` due to the way Documenter currently resolves `@ref`. A `@ref`d code literal will search for a docstring, rather than a header, to link to and the expression `stdlib-blas-trans` looks like a `-` method call to it.

I see whether it can be changed to act a bit more intelligently here.

andreasnoack commented on doc/src/stdlib/linalg.md:
```diff
@@ -157,6 +157,38 @@ linear algebra routines it is useful to call the BLAS functions directly.
 that overwrite one of the input arrays have names ending in `'!'`.  Usually, a BLAS function has
 four methods defined, for `Float64`, `Float32`, `Complex128`, and `Complex64` arrays.
 
+### [BLAS Character Arguments](@id stdlib-blas-chars)
```
Great addition.

kshyatt commented:
Ok so we've got the [`eigfact` for `SymTridiagonal](https://github.com/JuliaLang/julia/pull/19691#discussion_r94042166) issue and the [linking to the character table](https://github.com/JuliaLang/julia/pull/19691#discussion_r94005194) issue. Should I just leave this open until those are both fixed? What is actionable, by me, at this point?

andreasnoack commented:
I don't think this should wait for on a fix for the issues with truncated eigendecompositions. That can be tracked in a separate issue.

tkelman commented:
can at least document what's different about the returned factorization object in the current code

tkelman commented on base/linalg/symmetric.jl:
```diff
@@ -278,22 +278,116 @@ eigfact{T1<:Real,T2}(A::RealHermSymComplexHerm{T1,T2}) = (T = eltype(A); S = pro
+!!! note
+    If `irange` is not `1:n`, where `n` is the dimension of `A`, then the returned factorization
+    will be a *truncated* factorization and in this case, the determinant may be 0 and other methods
+    (such as [`inv`](@ref)) called on the factorization may throw.
```
it sounds like this is what we'd like to do, but isn't the case yet? it might give nonsense answers at the moment?

andreasnoack replied:
Just state that it is truncated for now. Right now the results of `inv` and `det` are wrong when the factorization is truncated.

kshyatt replied:
Gotcha, thanks!

kshyatt commented:
If I fix the merge conflict and this passes CI, is it ready to merge?

tkelman commented:
if @MichaelHatherly can confirm that updating the Documenter version fixes the table cross-reference issue

MichaelHatherly commented:
Yes, should be fine now with 0.8.7.
